Framework/Algorithms/src/SofQCommon.cpp

```diff
@@ -82,5 +85,142 @@ double SofQCommon::getEFixed(const Geometry::IDetector &det) const {
+  auto min = std::numeric_limits<double>::max();
+  auto max = std::numeric_limits<double>::lowest();
+  for (size_t i = 0; i < ws.getNumberHistograms(); ++i) {
+    const auto &x = ws.x(i);
+    if (x.front() < min) {
+      min = x.front();
+    }
+    if (x.back() > max) {
+      max = x.back();
+    }
+  }
+  return std::make_pair(min, max);
+}
+
+/**
+ * Return a pair of (minimum Q, maximum Q) for given workspace.
+ * @param ws a workspace
+ * @param minE minimum energy transfer in ws
+ * @param maxE maximum energy transfer in ws
+ * @return a pair containing global minimun and maximum Q
+ */
+std::pair<double, double> SofQCommon::qBinHints(const API::MatrixWorkspace &ws,
+                                                const double minE,
+                                                const double maxE) const {
+  if (m_emode == 1) {
+    return qBinHintsDirect(ws, minE, maxE);
+  }
+  return qBinHintsIndirect(ws, minE, maxE);
+}
+
+/**
+ * Return a pair of (minimum Q, maximum Q) for given
+ * direct geometry workspace.
+ * @param ws a workspace
+ * @param minE minimum energy transfer in ws
+ * @param maxE maximum energy transfer in ws
+ * @return a pair containing global minimun and maximum Q
+ */
+std::pair<double, double>
+SofQCommon::qBinHintsDirect(const API::MatrixWorkspace &ws, const double minE,
+                            const double maxE) const {
+  using namespace Mantid::PhysicalConstants;
+  auto minTheta = std::numeric_limits<double>::max();
+  auto maxTheta = std::numeric_limits<double>::lowest();
+  const auto &spectrumInfo = ws.spectrumInfo();
+  for (size_t i = 0; i < spectrumInfo.size(); ++i) {
+    if (spectrumInfo.isMasked(i) || spectrumInfo.isMonitor(i)) {
+      continue;
+    }
+    const auto theta = spectrumInfo.twoTheta(i);
+    if (theta < minTheta) {
+      minTheta = theta;
+    }
+    if (theta > maxTheta) {
+      maxTheta = theta;
+    }
+  }
+  if (minTheta == std::numeric_limits<double>::max()) {
+    throw std::runtime_error("Could not determine Q binning: workspace does "
+                             "not contain usable spectra.");
+  }
+  const auto incidentKSq = m_efixed / E_mev_toNeutronWavenumberSq;
+  const auto incidentK = std::sqrt(incidentKSq);
+  const auto minEnergy = m_efixed - maxE;
+  const auto maxEnergy = m_efixed - minE;
+  const auto minKSq = minEnergy / E_mev_toNeutronWavenumberSq;
+  const auto minK = std::sqrt(minKSq);
+  const auto maxKSq = maxEnergy / E_mev_toNeutronWavenumberSq;
+  const auto maxK = std::sqrt(maxKSq);
+  std::array<double, 4> q;
+  q[0] = std::sqrt(incidentKSq + minKSq -
+                   2. * incidentK * minK * std::cos(minTheta));
+  q[1] = std::sqrt(incidentKSq + minKSq -
+                   2. * incidentK * minK * std::cos(maxTheta));
+  q[2] = std::sqrt(incidentKSq + maxKSq -
+                   2. * incidentK * maxK * std::cos(minTheta));
+  q[3] = std::sqrt(incidentKSq + maxKSq -
+                   2. * incidentK * maxK * std::cos(maxTheta));
```

Comment:
It feels like this conversion must be something that has been implemented before somewhere else?